components/tidb_query/src/codec/mysql/json/mod.rs

```diff
@@ -32,38 +32,240 @@ use crate::codec::data_type::{Decimal, Real};
 use crate::codec::mysql;
 use crate::codec::mysql::{Duration, Time, TimeType};
 use crate::expr::EvalContext;
+use codec::number::{NumberCodec, F64_SIZE, I64_SIZE};
+use constants::{JSON_LITERAL_FALSE, JSON_LITERAL_NIL, JSON_LITERAL_TRUE};
 
 const ERR_CONVERT_FAILED: &str = "Can not covert from ";
 
-/// Json implements type json used in tikv, it specifies the following
-/// implementations:
-/// 1. Serialize `json` values into binary representation, and reading values
-///  back from the binary representation.
-/// 2. Serialize `json` values into readable string representation, and reading
-/// values back from string representation.
-/// 3. sql functions like `JSON_TYPE`, etc
+/// The types of `Json` which follows https://tools.ietf.org/html/rfc7159#section-3
+#[derive(Eq, PartialEq, FromPrimitive, Clone, Debug, Copy)]
+pub enum JsonType {
+    Object = 0x01,
+    Array = 0x03,
+    Literal = 0x04,
+    I64 = 0x09,
+    U64 = 0x0a,
+    Double = 0x0b,
+    String = 0x0c,
+}
+
+impl From<u8> for JsonType {
+    fn from(src: u8) -> JsonType {
+        num::FromPrimitive::from_u8(src).unwrap()
+    }
+}
+
+/// Represents a reference of JSON value aiming to reduce memory copy.
+#[derive(Clone, Debug)]
+pub struct JsonRef<'a> {
+    type_code: JsonType,
+    // Referred value
+    value: &'a [u8],
+}
+
+impl<'a> JsonRef<'a> {
+    pub fn new<T: Into<JsonType>>(type_code: T, value: &[u8]) -> JsonRef<'_> {
+        JsonRef {
+            type_code: type_code.into(),
+            value,
+        }
+    }
+
+    /// Returns an owned Json via copying
+    pub fn to_owned(&self) -> Json {
+        Json {
+            type_code: self.type_code,
+            value: self.value.to_owned(),
+        }
+    }
+
+    /// Returns the JSON type
+    pub fn get_type(&self) -> JsonType {
+        self.type_code
+    }
+
+    /// Returns the underlying value slice
+    pub fn value(&self) -> &'a [u8] {
+        &self.value
+    }
+
+    /// Returns the JSON value as u64
+    pub fn get_u64(&self) -> u64 {
+        NumberCodec::decode_u64_le(self.value())
```

Comment:
no need to check type code?

Reply:
The byte for the JSON type is not included in `self.value()`

Reply:
How about checking `self.type_code`?